### PR TITLE
Decrypt and Import full backups in chunk with progress

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes https://github.com/element-hq/element-web/issues/26856
Fixes https://github.com/element-hq/element-web/issues/23359 as it will chunk the importing of keys.

## Current State
Currently Web will download all the keys from the server, then will decrypt them all and keep them in memory (with no feedbacks during the decryption process), then will import them all at once to the crypto store.

## Changes

Now the keys will be split in chunks, then each chunk is decrypted and imported before processing to the next one.

restore_progress

This is a good improvement, but more work need to be done to have a proper download of a big backup. As an example you can't cancel an import at the moment. So if you close the dialog it will continue without feedback, and you can also start a new one in parallel.

## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Decrypt and Import full backups in chunk with progress ([\#4005](https://github.com/matrix-org/matrix-js-sdk/pull/4005)). Fixes element-hq/element-web#26856 and element-hq/element-web#23359.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -3970,15 +4013,89 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+     * It will decrypt the keys in chunks and call the `block` callback for each chunk.
+     *
+     * @param res - The response from the server containing the keys to be decrypted.
+     * @param backupDecryptor - An instance of the BackupDecryptor class used to decrypt the keys.
+     * @param chunkSize - The size of the chunks to be processed at a time.
+     * @param block - A callback function that is called for each chunk of keys.
+     *
+     * @returns A promise that resolves when the decryption is complete.
+     */
+    private async handleDecryptionOfAFullBackup(
```
Comment (dbkr) [inline src/client.ts:4051]: This could be a nice use of a generator function, potentially? Not a blocker though.

**Review by dbkr — APPROVED** [review]
LGTM generally. I assume this still downloads the lot in one lump, but this would need server changes so one step at a time and this is definitely a step in the right direcvtion. Might want a look from someone crypto-y too.

File: src/client.ts [hunk 2]
```diff
@@ -3970,15 +4013,89 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+        block: (chunk: IMegolmSessionData[]) => Promise<void>,
+    ): Promise<void> {
+        const rooms = (res as IRoomsKeysResponse).rooms;
+
+        let groupChunkCount = 0;
+        let chunkGroupByRoom: Map<string, IKeyBackupRoomSessions> = new Map();
+
+        const handleChunkCallback = async (roomChunks: Map<string, IKeyBackupRoomSessions>): Promise<void> => {
+            const currentChunk: IMegolmSessionData[] = [];
+            for (const roomId of roomChunks.keys()) {
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply src/client.ts:4064]: No, ignore me - we want the `roomId` later, and `Map.entries()` appears to have an annoying interface

File: spec/integ/crypto/megolm-backup.spec.ts [hunk 3]
```diff
@@ -340,6 +340,108 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("megolm-keys backup (%s)", (backe
...
+            await aliceClient.startClient();
+
+            // tell Alice to trust the dummy device that signed the backup
+            await waitForDeviceList();
+            await aliceCrypto.setDeviceVerified(testData.TEST_USER_ID, testData.TEST_DEVICE_ID);
+
+            const fullBackup: {
+                rooms: {
+                    [roomId: string]: {
+                        [sessionId: string]: any;
```
Comment (andybalaam) [inline spec/integ/crypto/megolm-backup.spec.ts:363]: Do we need `any` here? Isn't it a string?

```suggestion
                        [sessionId: string]: string;
```

File: src/client.ts [hunk 4]
```diff
@@ -3970,15 +4013,89 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+            }
+            await block(currentChunk);
+        };
+
+        for (const [roomId, roomData] of Object.entries(rooms)) {
+            if (!roomData.sessions) continue;
+
+            chunkGroupByRoom.set(roomId, {});
+
+            for (const sessions of Object.entries(roomData.sessions)) {
```
Comment (andybalaam) [inline src/client.ts:4080]: Can we do something like this?

```suggestion
            for (const [sessionId, sessions] of Object.entries(roomData.sessions)) {
```

I found the `sessions[0]` stuff below confusing.

(I'm not sure whether `sessionId` and `sessions` are the right names... as I said I am confused ;-)

**Review by andybalaam — CHANGES_REQUESTED** [review]
Looks good - lots of tiny comments but nothing major. Thanks!

File: src/client.ts (hunk 4 above)
Comment (BillCarsonFr) [reply src/client.ts]: yes, fixed https://github.com/matrix-org/matrix-js-sdk/pull/4005/commits/62a1f9ea99ab2132587467205655ead71e9d0fd3

File: src/client.ts [hunk 5]
```diff
@@ -3970,15 +4013,89 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+     * This method calculates the total number of keys present in the response of a `/room_keys/keys` call.
+     *
+     * @param res - The response from the server containing the keys to be counted.
+     *
+     * @returns The total number of keys in the backup.
+     */
+    private getTotalKeyCount(res: IRoomsKeysResponse): number {
+        const rooms = res.rooms;
+        let totalKeyCount = 0;
+        for (const entry of Object.entries(rooms)) {
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/client.ts]: yes, fixed https://github.com/matrix-org/matrix-js-sdk/pull/4005/commits/62a1f9ea99ab2132587467205655ead71e9d0fd3

File: spec/integ/crypto/megolm-backup.spec.ts (hunk 3 above)
Comment (BillCarsonFr) [reply spec/integ/crypto/megolm-backup.spec.ts:363]: good catch, I messed up the type. It should be
```
                    [roomId: string]: {
                        sessions: {
                            [sessionId: string]: KeyBackupSession;
                        };
                    };

```

File: spec/integ/crypto/megolm-backup.spec.ts [hunk 6]
```diff
@@ -340,6 +340,108 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("megolm-keys backup (%s)", (backe
...
+            const expectedTotal = keysPerRoom.reduce((a, b) => a + b, 0);
+            for (let i = 0; i < keysPerRoom.length; i++) {
+                const roomId = `!room${i}:example.com`;
+                fullBackup.rooms[roomId] = { sessions: {} };
+                for (let j = 0; j < keysPerRoom[i]; j++) {
+                    const sessionId = `session${j}`;
+                    // put the same fake session data, not important for that tet
+                    fullBackup.rooms[roomId].sessions[sessionId] = testData.CURVE25519_KEY_BACKUP_DATA;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply spec/integ/crypto/megolm-backup.spec.ts]: extracted function `createBackupDownloadResponse` https://github.com/matrix-org/matrix-js-sdk/pull/4005/commits/722291183c4eac5f365995e028a79fc76aa6cca0

File: spec/integ/crypto/megolm-backup.spec.ts [hunk 7]
```diff
@@ -340,6 +340,108 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("megolm-keys backup (%s)", (backe
...
+                stage: "load_keys",
+                failures: 0,
+            });
+
+            expect(progressCallback).toHaveBeenCalledWith({
+                total: expectedTotal,
+                successes: 832,
+                stage: "load_keys",
+                failures: 0,
+            });
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply spec/integ/crypto/megolm-backup.spec.ts:434]: There is already a `expect(importSpy).toHaveBeenCalledTimes(5);`

Anyhow added more checks and cleaned a bit the test https://github.com/matrix-org/matrix-js-sdk/pull/4005/commits/628bd861b5afe39bac7c16da468c303c27717179

File: spec/integ/crypto/megolm-backup.spec.ts [hunk 8]
```diff
@@ -340,6 +340,108 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("megolm-keys backup (%s)", (backe
...
+                },
+            );
+
+            expect(result.imported).toStrictEqual(expectedTotal);
+            // should be called 5 times: 200*4 plus one chunk with the remaining 32
+            expect(importSpy).toHaveBeenCalledTimes(5);
+
+            expect(progressCallback).toHaveBeenCalledWith({
+                // unfortunately there is no proper enum for stages :/
+                // react sdk expect some values though
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply spec/integ/crypto/megolm-backup.spec.ts]: Removed the comment as there is already a TODO in the code https://github.com/matrix-org/matrix-js-sdk/blob/c8855426282aa0f59d446893c93e0a4f2859b537/src/crypto-api.ts#L588

File: src/client.ts [hunk 9]
```diff
@@ -3934,25 +3934,63 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+                                progressCallback({
+                                    total: totalKeyCount,
+                                    successes: decryptedKeyCount,
+                                    stage: "load_keys",
+                                    failures: 0,
+                                });
+                            }
+                        } catch (e) {
+                            // We failed to import some keys, but we should still try to import the rest?
+                            // Log the error and continue
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/client.ts:3965]: A fix and new tests to check decryption failure and import failures https://github.com/matrix-org/matrix-js-sdk/pull/4005/commits/c414af3e40fcef5510b9c8d8457419ef0d6c4aff

File: spec/integ/crypto/megolm-backup.spec.ts [hunk 10]
```diff
@@ -340,17 +354,179 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("megolm-keys backup (%s)", (backe
...
+                failures: 0,
+            });
+        });
+
+        it("Should continue to process backup if a chunk import fails and report failures", async function () {
+            // @ts-ignore - mock a private method for testing purpose
+            aliceCrypto.importBackedUpRoomKeys = jest
+                .fn()
+                .mockImplementationOnce(() => {
+                    // fail to import fist chunk
```
Comment (andybalaam) [inline spec/integ/crypto/megolm-backup.spec.ts:444]:
```suggestion
                    // Fail to import first chunk
```

**Review by andybalaam — APPROVED** [review]
Looks great - one typo. Thanks!

**Pull request merged**